### PR TITLE
Add range indexing to UniformScaling

### PR DESCRIPTION
**Pull request by dalum**
Ref: JuliaLang/LinearAlgebra.jl#454. I thought this might be a usable alternative to `eye` and having callable `UniformScaling`.

### CODE REVIEW

File: base/linalg/uniformscaling.jl [hunk 1]
```diff
@@ -53,6 +53,29 @@ eltype(::Type{UniformScaling{T}}) where {T} = T
 ndims(J::UniformScaling) = 2
 getindex(J::UniformScaling, i::Integer,j::Integer) = ifelse(i==j,J.λ,zero(J.λ))
 
+function getindex(x::UniformScaling{T}, n::AbstractRange{<:Integer}, m::AbstractRange{<:Integer}) where {T}
+    if length(n) == length(m) && step(n) == step(m)
+        k = start(n) - start(m)
+        v = collect(Iterators.repeated(x.λ, length(n) - abs(k)))
```
Comment (stevengj) [inline base/linalg/uniformscaling.jl]: Probably better to use `fill` here.

**Comment by stevengj** [comment]
It's kind of unsatisfying to have this a type-unstable `getindex`.   I'm not sure I would encourage people to use an API like this.

File: base/linalg/uniformscaling.jl [hunk 2]
```diff
@@ -53,6 +53,29 @@ eltype(::Type{UniformScaling{T}}) where {T} = T
 ndims(J::UniformScaling) = 2
 getindex(J::UniformScaling, i::Integer,j::Integer) = ifelse(i==j,J.λ,zero(J.λ))
 
+function getindex(x::UniformScaling{T}, n::AbstractRange{<:Integer}, m::AbstractRange{<:Integer}) where {T}
+    if length(n) == length(m) && step(n) == step(m)
+        k = start(n) - start(m)
```
Comment (mbauman) [inline base/linalg/uniformscaling.jl]: These should be `first`, not `start`

File: base/linalg/uniformscaling.jl [hunk 3]
```diff
@@ -53,6 +53,29 @@ eltype(::Type{UniformScaling{T}}) where {T} = T
...
+        v = fill(x.λ, length(n) - abs(k))
```
Comment (mbauman) [inline base/linalg/uniformscaling.jl]: Maybe move this into the `length(v) > 0` branch since it's only used there.

**Review by mbauman — DISMISSED** [review]
Just a few minor changes.  :+1:

File: base/linalg/uniformscaling.jl (hunk 3 above)
Comment (mbauman) [reply base/linalg/uniformscaling.jl]: Of course, you'll need to change that condition to `length(n) - abs(k) > 0`.

**Comment by mbauman** [comment]
Interestingly, this should also work for negative indices.  At first I thought about requesting a bounds check there, but given that `I` should work similarly well for OffsetArrays… as should this definition as-is… I think we could leave that ability as a feature.  Thoughts?

File: base/linalg/uniformscaling.jl [hunk 4]
```diff
@@ -53,6 +53,29 @@ eltype(::Type{UniformScaling{T}}) where {T} = T
 ndims(J::UniformScaling) = 2
 getindex(J::UniformScaling, i::Integer,j::Integer) = ifelse(i==j,J.λ,zero(J.λ))
 
+function getindex(x::UniformScaling{T}, n::AbstractRange{<:Integer}, m::AbstractRange{<:Integer}) where T
+    if length(n) == length(m) && step(n) == step(m)
+        k = first(n) - first(m)
+        if length(n) - abs(k) > 0
+            v = fill(x.λ, length(n) - abs(k))
+            return spdiagm(k => v)
```
Comment (stevengj) [inline base/linalg/uniformscaling.jl]: This doesn't seem right for non-unit steps.  Consider
```jl
julia> eye(10,10)[1:2:7, 2:2:8]
4×4 Array{Float64,2}:
 0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0
```

Comment (eveydee) [reply base/linalg/uniformscaling.jl]: Nice catch. Should be fixed by additionally checking if `k % step(n) == 0`.

**Comment by eveydee** [comment]
@mbauman That it works for negative indices was intentional, since integer indexing works that way. Any other infinite-dimensional abstract array could also allow arbitrary integer indexing, so I see no good argument for enforcing strictly positive indexing.

**Comment by stevengj** [comment]
Do we want to resurrect this at some point?

It occurred to me recently that it might be nice to have the related syntax `I[1:n, n]` for forming the n-th unit vector (as a `Vector`).

**Comment by dalum** [comment]
@stevengj I updated the PR to include the indexing you suggested.  I also had to change the behaviour of range indexing to return a dense array, since the separation of LinearAlgebra and SparseArrays meant that sparse functions were not available in the scope.  I think it's better not to make LinearAlgebra depend on SparseArrays for this functionality?

The implementation right now is about twice as fast as first instancing a diagonal identity matrix through `I(n)` and then subsequently indexing into it, at least for small matrices.

File: stdlib/LinearAlgebra/src/uniformscaling.jl [hunk 5]
```diff
@@ -78,6 +78,28 @@ ndims(J::UniformScaling) = 2
 Base.has_offset_axes(::UniformScaling) = false
 getindex(J::UniformScaling, i::Integer,j::Integer) = ifelse(i==j,J.λ,zero(J.λ))
 
+getindex(x::UniformScaling, n::Integer, m::AbstractRange{<:Integer}) = getindex(x, m, n)
```
Comment (stevengj) [inline stdlib/LinearAlgebra/src/uniformscaling.jl:81]: Should this be `getindex(x, m, n)'` to transpose?

Comment (stevengj) [reply stdlib/LinearAlgebra/src/uniformscaling.jl:81]: No, I guess not.

**Review by stevengj — APPROVED** [review]

**Comment by stevengj** [comment]
Needs a NEWS item?

**Comment by dalum** [comment]
It is not entirely clear to me where to add the compat annotation.  There are no existing docstrings for `getindex(::UniformScaling)`, so should I add one as well?  I'm just a little worried about cluttering `?getindex`.

**Comment by mbauman** [comment]
Ah, very good question.  I agree about it not really belonging in `getindex`, but I think we could say something in the type itself:

https://github.com/JuliaLang/julia/blob/147d5b101cea8c845b81c80943a8089bd59f5e1a/stdlib/LinearAlgebra/src/uniformscaling.jl#L10

It could be nice to say something about it acting somewhat matrix-like with some indexing supported but without an explicit `size` — and then a compat entry would fit there.

**Comment by dalum** [comment]
Thanks! Not sure if I worded it right, but I added a docstring example that should give some idea about the kind of indexing that is supported.

**Comment by mbauman** [comment]
Looks great to me, thank you!

**Review by mbauman — APPROVED** [review]

**Review by stevengj — APPROVED** [review]

**Comment by stevengj** [comment]
Should we make `I(m,n)` return `I[1:m, 1:n]`?

**Comment by StefanKarpinski** [comment]
Seems like the only sane definition to me, so 👍

**Pull request merged**